### ISSUE TITLE
Weird bug with default server directive that has both HTTP and HTTPS?

### ISSUE THREAD

**Issue by p0358** [caddyserver/caddy#4829]
With the following Caddyfile:
```
http:// {
        respond "DEFAULT"
}
```
Caddy will respond to a HTTP request on port 80 over direct IP as expected with "DEFAULT".
```
> GET / HTTP/1.1
> Host: 192.168.1.100
> User-Agent: curl/7.79.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: Caddy
< Date: Tue, 07 Jun 2022 23:47:19 GMT
< Content-Length: 7
<
DEFAULT
```

However, if we try adding a https:// there like so:
```
http:// https:// {
        respond "DEFAULT"
}
```
suddenly the HTTP port starts redirecting to HTTPS.
```
> GET / HTTP/1.1
> Host: 192.168.1.100
> User-Agent: curl/7.79.1
> Accept: */*
>
< HTTP/1.1 308 Permanent Redirect
< Connection: close
< Location: https://192.168.1.100/
< Server: Caddy
< Date: Tue, 07 Jun 2022 23:46:06 GMT
< Content-Length: 0
<
```
This is most unexpected, given Caddy is normally supposed to support mixed HTTP and HTTPS site blocks, like this:
```
http://test.example.com test.example.com {
        respond "DOMAIN"
}
```
In the above example, you can normally access test.example.com over HTTP as expected (over HTTPS too ofc).

What's even more surprising, is that when we modify the second Caddyfile snippet to be this:
```
test.example.com {
        # must be https
        respond "TEST"
}

http:// https:// {
        respond "DEFAULT"
}
```
**Suddenly HTTP port responds with "DEFAULT" as expected again!**

Looking further at the default server example though, it seems that its HTTPS port does not work (no matter if tried by IP or dummy domain):
```
>curl -k -v https://192.168.1.100
*   Trying 192.168.1.100:443...
* Connected to 192.168.1.100 (192.168.1.100) port 443 (#0)
* schannel: disabled automatic use of client certificate
* schannel: using IP address, SNI is not supported by OS.
* schannel: ALPN, offering http/1.1
* schannel: next InitializeSecurityContext failed: SEC_E_ILLEGAL_MESSAGE (0x80090326) - This error usually occurs when a fatal SSL/TLS alert is received (e.g. handshake failed). More detail may be available in the Windows System event log.
* Closing connection 0
curl: (35) schannel: next InitializeSecurityContext failed: SEC_E_ILLEGAL_MESSAGE (0x80090326) - This error usually occurs when a fatal SSL/TLS alert is received (e.g. handshake failed). More detail may be available in the Windows System event log.
```
If I add `tls internal` into the "DEFAULT" block, nothing changes. If I add `tls <path>.pem <path>.pem` manually with a path to the certificate into the section, Caddy fails to start:
```
run: adapting config using caddyfile: server listening on [:80] is HTTP, but attempts to configure TLS connection policies
```

---

# **So to summarize, it seems there are two issues**:
1. Inconsistent behavior of the "default" (no-hostname) site block for both HTTP and HTTPS based on existence of **another** unrelated HTTPS site block (if "default" block is the only one, HTTP is redirected to HTTPS for no reason?).
2. You are supposed to be able to create a site block that can be served over HTTP and HTTPS both, but you seemingly cannot then use `tls` directive inside to set the "internal" or custom certificate paths, it only works with default encryption that specifies no custom options. Expected behavior would be that whatever is set in `tls` option would simply be ignored for HTTP connections...

It is worth to note that I was previously successfully using both Apache2 and nginx with such configuration that is basically the default if you enable SSL there, that is the same default server block is also served over HTTPS with a self-signed certificate, which works both over IP or any unrecognized domain, thus titled the "default".

### Workaround attempt

I tried this:
```
(default) {
        respond "DEFAULT"
}

http:// {
        import default
}

https:// {
        # I could not access this over HTTPS with neither TLS options or lack thereof
        #tls internal
        # adding the line below (instead of the above ofc) makes the Caddy server start with no error, but stop responding to ANY requests, both HTTP or HTTPS (???)
        #tls /etc/ssl/certs/ssl-cert-snakeoil.pem /etc/ssl/private/ssl-cert-snakeoil.key
        import default
}
```
Result: I still cannot load the server via HTTP, I still need to add a random HTTPS site block in order to be able to reach the default HTTP site.

---

Now of course I could simply accept that I will not get an unsecure default unnamed HTTPS endpoint with Caddy that I could by default with Apache2 and nginx. But I think my attempts still bring up some logic errors in Caddy config processing that shouldn't be behaving the way they behave regardless.

---

Caddy version: v2.5.1 h1:bAWwslD1jNeCzDa+jDCNwb8M3UJ2tPa8UZFFzPVmGKs=

**Comment by p0358** [comment]
It seems that even an example of
```
:80 http://domain.example.com domain.example.com
```
confuses Caddy.

With a site block like above you can access the server via IP and port 80 as expected, but trying to access it via `http://domain.example.com` causes the unexpected redirect to HTTPS version to occur

**Comment by francislavoie** [comment]
This is working as intended, actually. The HTTP->HTTPS redirect routes get inserted _just before_ any HTTP catch-all routes, so that it's possible to keep the redirect and force certain domains to be served over HTTP.

See https://github.com/caddyserver/caddy/pull/4033 and https://github.com/caddyserver/caddy/issues/3212 which explain the problem.

To work around it in your situation, you'll want to use the global option `auto_https disable_redirects` which will remove the redirect routes and let HTTP requests actually reach your catch-all block.

**Comment by p0358** [comment]
> To work around it in your situation, you'll want to use the global option auto_https disable_redirects which will remove the redirect routes and let HTTP requests actually reach your catch-all block.

Honestly I think this might be the best solution in my case, thanks.

> As for the `tls` directive thing, I disagree. It's a configuration error to use tls inside of a site block configuring an HTTP server. Split them up.

I think this is inconsistent though -- because a single site block for HTTP and HTTPS is otherwise allowed and perfectly functional. It works fine with automatic HTTPS, but if you want to define custom TLS options, then you are required to split them up just to insert the `tls` option. I think that's just pretty inconvenient.

---

I still cannot wrap my head around on how to get catch-all HTTPS working, is it even supported? And speaking of which, I found another weird behavior. Suppose this config file:
```
{
    auto_https disable_redirects
}
(default) {
    respond "DEFAULT"
}
http:// {
    import default
}
https:// {
    tls internal
    import default
}
# ...
domain1.example.com:8443 {
    respond "UNRELATED BLOCK"
}
```
Let's ignore the last block that works independently and let's assume *.example.com resolves to our server's IP. The following happens:
1. I can access catch-all http://<serverIP> and http://domain1.example.com and http://random.example.com, all lead to "DEFAULT"
2. I cannot access catch-all https://<serverIP> or https://random.example.com
3. I can access only https://domain.example.com

Note that in point 3 I can access the catch-all only with a certificate that's in Caddy's memory, despite being there (generated by Let's Encrypt) for a completely unrelated block

Now if I do this:
```
{
    auto_https disable_redirects
}
(default) {
    respond "DEFAULT"
}
http:// {
    import default
}
https:// {
    # self-signed certs
    tls /var/lib/caddy/ssl-cert-snakeoil.pem /var/lib/caddy/ssl-cert-snakeoil.key
    import default
}
```

It works, I can curl the server on HTTPS with any domain or IP! But!!! Apparently now I cannot add any normal HTTPS site block on a normal. Let's add this:
```
domain1.example.com {
    respond "NON DEFAULT"
}
```
**If I open https://domain1.example.com, I get "NON DEFAULT" served, which means the block works. But I get certificate specified in https:// catch-all instead!!!** This is clearly a bug in certificate selection routines, since otherwise the block wouldn't be used at all if catch-all took precedence.

The automatic certificate in my case of `domain1.example.com` exists, and I can add a site block on this domain on another port and it works. It also works on the default port if I comment out the catch-all https block...

**Comment by francislavoie** [comment]
> I think this is inconsistent though -- because a single site block for HTTP and HTTPS is otherwise allowed and perfectly functional. It works fine with automatic HTTPS, but if you want to define custom TLS options, then you are required to split them up just to insert the tls option. I think that's just pretty inconvenient.

I don't think it's inconsistent. It's the difference between implicit and explicit config for HTTPS. If you let Caddy enable automatic HTTPS (implicitly) then a single block is fine. Obviously you need to disable redirects if you want to do that though. But if you need to configure `tls` explicitly, then yes you need to split it up, because otherwise it's a logic error; you can't configure TLS for HTTP.

> I still cannot wrap my head around on how to get catch-all HTTPS working, is it even supported?

Yes it is supported, but you need to have a certificate to serve for the requested domain. If you don't put a domain in your config, Caddy won't be able to issue a certificate ahead of time, so it won't be able to complete a handshake.

If you use `tls internal`, then you are telling Caddy to use its internal CA, but you still didn't tell it which domains to manage a cert for. If you don't know the domains ahead of time, then you need to use On-Demand TLS, which enables a mode where Caddy will attempt to issue a certificate for the domain that was requested in the TLS handshake (the SNI field). But keep in mind, enabling `on_demand` has risk, because it's unbounded by default; some attacker could force your server to issue an infinite amount of TLS certificates and fill up your disk storage with garbage, if your server is publicly accessible.

> Note that in point 3 I can access the catch-all only with a certificate that's in Caddy's memory, despite being there (generated by Let's Encrypt) for a completely unrelated block

Yeah, that's normal. Under the hood, Caddy's actual config language is JSON. If you run `caddy adapt --pretty` on your Caddyfile, you'll see what your config looks like. It'll probably be a bit clearer what's going on.

Keep in mind that this won't show the routes that automatic HTTPS adds in, because that happens at runtime just before starting the server, it's not included in the config.

How it works is Caddy has a `tls` app, which is what does all the certificate maintenance and caching, and there's the `http` app, which is what actually runs the server. The HTTP app asks the TLS app "hey do you have a cert for _this domain_?" and if the TLS app has one in cache, it'll use it to serve.

In JSON config, there's no concept of "sites" at all, that's just syntax sugar in the Caddyfile to try to make it easier to write configs. But you're kinda trying to abuse all the edgecases here, so of course you'll run up against all pitfalls/caveats of that abstraction.

> If I open [domain1.example.com](https://domain1.example.com/), I get "NON DEFAULT" served, which means the block works. But I get certificate specified in https:// catch-all instead!!!

In that case, you would need to use `auto_https ignore_loaded_certs` to tell Caddy "don't use certs from `tls` loaded into cache, and try to manage a cert anyways if there's an explicit domain listed in the config". But we didn't make it possible to specify both `disable_redirects` and `ignore_loaded_certs` at the same time... because that's _insanely_ edgecase-y. Do you _really_ need to do that? It seems absolutely crazy-bonkers to me. If you really, really need that, honestly you should probably just use JSON config directly, so you have full control over what's going on.

**Comment by mholt** [comment]
I think there might be improvements we could make here, either in docs or in implementation, but I need to reread this issue with fresher eyes.

**Comment by whitestrake** [comment]
> I don't think it's inconsistent. It's the difference between implicit and explicit config for HTTPS.

I think there is inconsistency. To bring this down to, ultimately, the simplest expression I can demonstrate with:

```
https:// {
  tls internal {
    on_demand
  }
  respond "HTTPS"
}

http:// {
  respond "HTTP"
}
```

And I make a request to `http://localhost`, I expect a 200 OK with "HTTP", period. Currently, I get a redirect to `https://localhost` and receive a 200 OK with "HTTPS". I'm not sure I can get more explicit than this, but the HTTP listener is just getting completely trodden over right now.

This changes as soon as I change the HTTPS site to `https://localhost`, in which case its overriding redirect becomes limited in scope to `http://localhost` requests and not e.g. `http://local.test`.

I haven't investigated, but I have a hunch this is part of the interaction of your logic, @francislavoie (which allows for another site to override a HTTP catch-all with its listener), and the fact that Caddy seems to interpret an explicit HTTPS listener as "HTTPS with a HTTP redirect", meaning that this HTTPS catch-all is having its catch-all redirect listener override the _actually configured_ HTTP listener.

Does that sound like it makes sense?

That said, from an ergonomics perspective, I think that:
- A manually configured catch-all HTTP site should take precedence over an auto-configured HTTP->S catch-all redirect
- I shouldn't have to use `auto_https disable_redirects` in the long run for this to work

And further (possibly strictly beyond the scope of this issue, but related because I believe it's interacting here):
- An explicit HTTPS listener shouldn't configure a HTTP redirect (only scheme-agnostic sites should)

**Comment by francislavoie** [comment]
> I shouldn't have to use auto_https disable_redirects in the long run for this to work

I disagree, this is exactly the situation where you should use it. I don't see how a redirect would ever be correct with your intent as described here, so you should turn off redirects.

**Comment by whitestrake** [comment]
Yeah, sure, I agree that it functions as a workaround, but there's _no way_ that a HTTPS and a HTTP block, explicitly configured in the Caddyfile, should be interpreted as "please ignore my HTTP block entirely and redirect it to HTTPS" in the long run.

Whether there's an additional check that does this automatically or not, the intent behind this configuration is abundantly clear, and any logic that produces a result counter to that intent with that incredibly simple configuration is, in my opinion, not ideal.

**Comment by mholt** [comment]
(FYI, I'll chime in too when I'm back at the office tomorrow)

**Comment by mholt** [comment]
I think both sides have good, valid points:

- If redirects are not the intent, they can be disabled (Francis)
- Implicit rules should never override explicit configuration (Matthew)

The counterpoints are, respectively:

- Putting `http://` site in your config _is_ the intention to disable or override the implicit redirects (Matthew)
- The explicit configuration regarding redirects is `auto_https disable_redirects`, not `http://` (Francis)

I agree with @francislavoie that more magic is not needed, nor is it the answer.

I agree with @whitestrake that the current behavior is unintuitive; a site explicitly defined as `http://` should be served as HTTP, not HTTPS.

---

The reason this behavior was introduced was because its inverse was also confusing (#3212):

```
example.com {
	# ...
}

http:// {
	respond "HTTP"
}
```

Here, the author expects `http://example.com` to redirect to `https://example.com`, but wants all other traffic served over HTTP. In other words, they want the auto-redirects to come _BEFORE_ their explicit configuration _just for sites they've defined._ Of course, this does a disservice to those who want to serve HTTP in all cases, or who have catch-all cases like above.

The first-proposed solution was to do the redirect manually:

```
http://example.com {
	redir https://{hostport}{uri}
}
```

but this requires enumerating every domain that you have an HTTPS site for, a second time. And you have to know the domains.

In #3212, Francis also proposed a handler-based solution, where a matcher+directive could be used within the `http://` site to issue the redirect without having to enumerate all the other domains. I didn't love that solution though due to its internal complexity (and that the list of names could be very long) -- at the time we didn't have binary search implemented for the host matcher, but now we do. The OP also didn't love this solution though because it kind of requires knowing how Caddy's internals work, utilizing a special placeholder, etc.

So in the end we went with a solution that injects the implicit redirect logic _before_ the explicit configuration in some cases. Of course, this current implementation has a few problems:

- Sometimes that is not the user's intent. (They may want their explicit `http://` configuration to be effective.)
- Although `auto_https disable_redirects` can be used, this disables ALL automatic redirects, which may also be inconvenient. Sounds like they would only want to disable the redirects for domains they don't have HTTPS sites for (and the domains are unknown in some cases, like the current issue).
- Somewhat logically complex, including an edge case that had to be fixed in a later PR (#4033, then #4243)

Personally, I'd like to roll back the current fix for this, so that we don't override explicit configuration. And implement a new fix that can satisfy both valid use cases without being complicated.

As for what that is, I'm not sure yet, but I'm hoping to think on this or talk about it here.

**Comment by whitestrake** [comment]
This summary post is excellent, @mholt. Just a few very quick clarifications of my own stance:

> a site explicitly defined as http:// should be served as HTTP, not HTTPS.

It’s arguably a step worse than just that; the explicit HTTP site wasn’t being served on HTTPS, it wasn’t being served _at all._ With the configuration I posted above, `respond "HTTP"` was _never_ served. That entire site block might as well have not existed at all.

> Putting http:// site in your config is the intention to disable or override the implicit redirects (Matthew)

To expound on this, naturally the existence of literal catch-all `http://` implies that it should be logically _possible_ for at least some request that can be made should be routed this way!

---

I would rank which HTTP listener configuration should 'win' in terms of priority by way of how explicitly defined they are:

1. Explicit HTTP configured site block with a hostname
2. HTTP redirect listener for a scheme-agnostic site with a hostname
3. Catch-all HTTP site block (no hostname)
4. HTTP redirect listener for explicit HTTPS-only sites with hostnames, or for HTTPS catch-alls

@francislavoie's PR to handle this case (where we have two catch-alls, one for each scheme) solves, in my view, the question of 4 beating 3.

**Comment by mholt** [comment]
Great, I think so too! (I amended my comment to better reflect your stance.)

@p0358 I'm going to merge Francis' patch in #5051 and release it. When you have a chance would you give it a try?

**Comment by ledlamp** [comment]
> I don't think it's inconsistent. It's the difference between implicit and explicit config for HTTPS. If you let Caddy enable automatic HTTPS (implicitly) then a single block is fine. Obviously you need to disable redirects if you want to do that though. But if you need to configure `tls` explicitly, then yes you need to split it up, because otherwise it's a logic error; you can't configure TLS for HTTP.

Why not just simply ignore the tls settings for http so you don't have to make two copies of the configuration?

Whether a site has TLS or not can be explicitly defined by http:// or https:// in the site name, and sites can have multiple names with different protocols and even different ports. So I think it would make more sense if the `tls` directives simply had no effect on listeners that it is not applicable to, than to reject this configuration alltogether and require redundant configuration for what benefit? This is how I was expecting it to work.
